simulation.hpp

```diff
@@ -129,6 +129,75 @@ class Simulation
         }
     }
 
+    /** Moves hosts from one location to another
+     *
+     * @param infected Currently infected hosts
+     * @param susceptible Currently susceptible hosts
+     * @param mortality_tracker Hosts that are infected at a specific time step
+     * @param total_plants 
+     * @param step the current step of the simulation
+     * @param last_index the last index to not be used from movements
+     * @param movements a vector of ints with row_from, col_from, row_to, col_to, and num_hosts
+     * @param movement_schedule a vector matching movements with the step at which the movement from movements are applied
+     */
+    unsigned movement(IntegerRaster& infected,
+                      IntegerRaster& susceptible, 
+                      IntegerRaster& mortality_tracker,
+                      IntegerRaster& total_plants,
+                      unsigned step, unsigned last_index,
+                      const std::vector<std::vector<int> >& movements,
+                      std::vector<unsigned> movement_schedule)
+    {
+        for (unsigned i = last_index; i < movements.size(); i++) {
+            auto moved = movements[i];
+            unsigned move_schedule = movement_schedule[i];
+            if (move_schedule != step) {
+                return i;
+            }
+            int infected_moved = 0;
+            int susceptible_moved = 0;
+            int total_hosts_moved = 0;
+            double infected_mean = 0;
+            double inf_ratio = 0;
+            int row_from = moved[0];
+            int col_from = moved[1];
+            int row_to = moved[2];
+            int col_to = moved[3];
+            int hosts = moved[4];
+            if (infected(row_from, col_from) > 0 || susceptible(row_from, col_from)) {
+                inf_ratio = double(infected(row_from, col_from)) / double(susceptible(row_from, col_from));
+                infected_mean = infected(row_from, col_from) * inf_ratio;
+                std::poisson_distribution<int> distribution(infected_mean);
+                if (infected_mean > 0)
+                    infected_moved += distribution(generator_);
+                }
+            if ((hosts - infected_moved) > susceptible(row_from, col_from)) {
+                susceptible_moved = susceptible(row_from, col_from);
+            } else {
+                susceptible_moved = hosts - infected_moved;
```

Comment:
infected_moved (result of poisson) can be huge, so susceptible_moved can become negative and later `infected(row_to, col_to)` too large unless I'm missing some condition here. It seems like an additional case is missing to account for this state which is basically "all moved hosts are the infected ones." The infected_moved needs to be limited by infected(row_from, col_from). If I'm thinking about this correctly, this theoretically increases probability of having all moved host infected.

Perhaps you could also put all checks for data input first (like num_hosts <= total_host, num_hosts <= susceptible() + infected()) and then resolve the other issues as they appear resulting in correct move counts. Perhaps this way it would be easier to see if we account for all the special cases.

This is a clear candidate for precondition/postcondition/assert for total_plants == infected + susceptible, but I'm not too excited about adding these to the code when the only tool for it is the assert macro. However, this is something what a separate test in test_simulation.cpp could check for.

Reply:
I adjusted this piece to first check hosts moved aren't greater than total hosts and then do the infected and susceptible checks based on the output from the previous check and to ensure that the infected draw from the Poisson distribution wasn't larger than the infected in the cell.